### PR TITLE
Pytorch: 9.7 seq2seq

### PR DESCRIPTION
**Pull request by ypandya**
*Description of changes:*

Pytorch port for 9.7. Sequence to Sequence

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_recurrent-modern/seq2seq.md [hunk 1]
```diff
@@ -251,6 +418,32 @@ def predict_s2s_ch9(model, src_sentence, src_vocab, tgt_vocab, num_steps,
     return ' '.join(tgt_vocab.to_tokens(predict_tokens))
 ```
 
+```{.python .input}
+#@tab pytorch
+#@save
+def predict_s2s_ch9(model, src_sentence, src_vocab, tgt_vocab, num_steps,
+                    device):
+    src_tokens = src_vocab[src_sentence.lower().split(' ')]
+    enc_valid_len = torch.Tensor([len(src_tokens)], device=device)
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:427]:
```suggestion
    enc_valid_len = torch.tensor([len(src_tokens)], dtype=torch.float32,  device=device)
```

File: chapter_recurrent-modern/seq2seq.md [hunk 2]
```diff
@@ -251,6 +418,32 @@ def predict_s2s_ch9(model, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+    src_tokens = d2l.truncate_pad(src_tokens, num_steps, src_vocab['<pad>'])
+    enc_X = torch.Tensor(src_tokens, device=device).long()
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:429]:
```suggestion
    enc_X = torch.tensor(src_tokens, dtype=torch.long, device=device)
```

File: chapter_recurrent-modern/seq2seq.md [hunk 3]
```diff
@@ -251,6 +418,32 @@ def predict_s2s_ch9(model, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+    # Add the  batch size dimension
+    enc_outputs = model.encoder(torch.unsqueeze(enc_X, dim=0),
+                                enc_valid_len)
+    dec_state = model.decoder.init_state(enc_outputs, enc_valid_len)
+    dec_X = torch.unsqueeze(torch.Tensor([tgt_vocab['<bos>']], device=device).long(), dim=0)
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:434]:
```suggestion
    dec_X = torch.unsqueeze(torch.tensor([tgt_vocab['<bos>']], dtype=torch.long, device=device), dim=0)
```

File: chapter_recurrent-modern/seq2seq.md [hunk 4]
```diff
@@ -251,6 +418,32 @@ def predict_s2s_ch9(model, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+    predict_tokens = []
+    for _ in range(num_steps):
+        Y, dec_state = model.decoder(dec_X, dec_state)
+        # The token with highest score is used as the next timestep input
+        dec_X = Y.argmax(axis=2)
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:439]:
```suggestion
        dec_X = Y.argmax(dim=2)
```

File: d2l/torch.py [hunk 5]
```diff
@@ -881,6 +881,128 @@ def forward(self, enc_X, dec_X, *args):
...
+            l = loss(Y_hat, Y_label, Y_vlen)
+            l.sum().backward() # Making the loss scalar for backward()
+            d2l.grad_clipping(model, 1)
+            num_tokens = Y_vlen.sum()
+            optimizer.step()
+            with torch.no_grad():
+                metric.add(l.sum(), num_tokens)
+        if epoch % 10 == 0:
+            animator.add(epoch, (metric[0]/metric[1],))
+    print('loss %.3f, %d tokens/sec on %s ' % (
```
Comment (AnirudhDagar) [inline d2l/torch.py:978]: We are now shifting to `f strings` as standard for `prints`. Please update this :)

File: chapter_recurrent-modern/seq2seq.md [hunk 6]
```diff
@@ -74,9 +102,20 @@ output, state = encoder(X)
 output.shape
 ```
 
+```{.python .input}
+#@tab pytorch
+encoder = Seq2SeqEncoder(vocab_size=10, embed_size=8, num_hiddens=16,
+                         num_layers=2)
+encoder.eval()
+X = torch.zeros(4, 7).long()
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:110]:
```suggestion
X = d2l.zeros((4, 7), dtype=torch.long)
```
We are now making a few functions standard across all frameworks. See [here](https://github.com/d2l-ai/d2l-en/blob/28b0d4ce61c3f6f79e95f57d0be284c71453f9f4/d2l/torch.py#L35)

File: chapter_recurrent-modern/seq2seq.md [hunk 7]
```diff
@@ -139,13 +211,33 @@ X = np.array([[1, 2, 3], [4, 5, 6]])
...
 Apply to $n$-dim tensor $X$, it sets `X[i, len[i]:, :, ..., :] = 0`. In addition, we can specify the filling value such as $-1$ as shown below.
 
 ```{.python .input  n=8}
 X = np.ones((2, 3, 4))
 npx.sequence_mask(X, np.array([1, 2]), True, value=-1, axis=1)
 ```
 
+```{.python .input}
+#@tab pytorch
+X = torch.ones(2, 3, 4)
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:236]:
```suggestion
X = d2l.ones(2, 3, 4)
```
We are now making a few functions standard across all frameworks. See [here](https://github.com/d2l-ai/d2l-en/blob/28b0d4ce61c3f6f79e95f57d0be284c71453f9f4/d2l/torch.py#L35)

File: chapter_recurrent-modern/seq2seq.md [hunk 8]
```diff
@@ -117,6 +156,29 @@ class Seq2SeqDecoder(d2l.Decoder):
...
+        self.dense = nn.Linear(num_hiddens, vocab_size)
+
+    def init_state(self, enc_outputs, *args):
+        return enc_outputs[1]
+
+    def forward(self, X, state):
+        X = self.embedding(X).permute(1, 0, 2)
+        out, state = self.rnn(X, state)
+        # Make the batch to be the first dimension to simplify loss
+        # computation
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:177]:
```suggestion
        # Make the batch to be the first dimension to simplify loss computation
```

Comment (AnirudhDagar) [reply chapter_recurrent-modern/seq2seq.md:177]: Please also update the same for MXNet

File: chapter_recurrent-modern/seq2seq.md [hunk 9]
```diff
@@ -161,13 +253,35 @@ class MaskedSoftmaxCELoss(gluon.loss.SoftmaxCELoss):
...
 
 ```{.python .input  n=10}
 loss = MaskedSoftmaxCELoss()
 loss(np.ones((3, 4, 10)), np.ones((3, 4)), np.array([4, 2, 0]))
 ```
 
+```{.python .input}
+#@tab pytorch
+loss = MaskedSoftmaxCELoss()
+loss(torch.ones(3, 4, 10), torch.ones(3, 4).long(), torch.tensor([4, 2, 0]))
```
Comment (AnirudhDagar) [inline chapter_recurrent-modern/seq2seq.md:282]:
```suggestion
loss(d2l.ones(3, 4, 10), d2l.ones((3, 4), dtype=torch.long), torch.tensor([4, 2, 0]))
```

**Comment by mli** [comment]
Job d2l-en/PR-1119/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by mli** [comment]
Job d2l-en/PR-1119/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by mli** [comment]
Job d2l-en/PR-1119/6 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by mli** [comment]
Job d2l-en/PR-1119/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by mli** [comment]
Job d2l-en/PR-1119/8 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by mli** [comment]
Job d2l-en/PR-1119/9 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by ypandya** [comment]
@AnirudhDagar I've made some more edits according to your comments. Please check and let me know if the latest commit needs any changes.

File: d2l/torch.py (hunk 5 above)
Comment (ypandya) [reply d2l/torch.py]: Updated.

**Comment by mli** [comment]
Job d2l-en/PR-1119/11 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1119/

**Comment by AnirudhDagar** [comment]
Thanks @ypandya 
LGTM! Merging
